src/core/lib/event_engine/windows/grpc_polled_fd_windows.cc

```diff
@@ -0,0 +1,711 @@
+          grpc_core::StatusToString(error).c_str());
+    }
+  }
+  if (error.ok()) {
+    read_buf_ = grpc_slice_sub_no_ref(
+        read_buf_, 0, winsocket_->read_info()->result().bytes_transferred);
+    read_buf_has_data_ = true;
+  } else {
+    grpc_core::CSliceUnref(read_buf_);
+    read_buf_ = grpc_empty_slice();
+  }
+  GRPC_ARES_RESOLVER_TRACE_LOG(
+      "fd:|%s| OnIocpReadable finishing. read buf length now:|%d|", GetName(),
+      GRPC_SLICE_LENGTH(read_buf_));
+  ScheduleAndNullReadClosure(error);
+}
+
+void GrpcPolledFdFactoryWindows::GrpcPolledFdWindows::OnIocpWriteable() {
+  grpc_core::MutexLock lock(mu_);
+  GRPC_ARES_RESOLVER_TRACE_LOG("OnIocpWriteableInner. fd:|%s|", GetName());
+  GPR_ASSERT(socket_type_ == SOCK_STREAM);
+  absl::Status error;
+  if (winsocket_->write_info()->result().wsa_error != 0) {
+    error = GRPC_WSA_ERROR(winsocket_->write_info()->result().wsa_error,
+                           "OnIocpWriteableInner");
+    GRPC_ARES_RESOLVER_TRACE_LOG(
+        "fd:|%s| OnIocpWriteableInner. winsocket_->write_info.wsa_error "
+        "code:|%d| msg:|%s|",
+        GetName(), winsocket_->write_info()->result().wsa_error,
+        grpc_core::StatusToString(error).c_str());
+  }
+  GPR_ASSERT(tcp_write_state_ == WRITE_PENDING);
+  if (error.ok()) {
+    tcp_write_state_ = WRITE_WAITING_FOR_VERIFICATION_UPON_RETRY;
+    write_buf_ = grpc_slice_sub_no_ref(
+        write_buf_, 0, winsocket_->write_info()->result().bytes_transferred);
+    GRPC_ARES_RESOLVER_TRACE_LOG(
+        "fd:|%s| OnIocpWriteableInner. bytes transferred:%d", GetName(),
+        winsocket_->write_info()->result().bytes_transferred);
+  } else {
+    grpc_core::CSliceUnref(write_buf_);
+    write_buf_ = grpc_empty_slice();
+  }
+  ScheduleAndNullWriteClosure(error);
+}
+
+grpc_slice GrpcPolledFdFactoryWindows::GrpcPolledFdWindows::FlattenIovec(
+    const struct iovec* iov, int iov_count) {
+  int total = 0;
+  for (int i = 0; i < iov_count; i++) {
+    total += iov[i].iov_len;
+  }
+  grpc_slice out = GRPC_SLICE_MALLOC(total);
+  size_t cur = 0;
+  for (int i = 0; i < iov_count; i++) {
+    for (size_t k = 0; k < iov[i].iov_len; k++) {
+      GRPC_SLICE_START_PTR(out)
+      [cur++] = (static_cast<char*>(iov[i].iov_base))[k];
+    }
+  }
+  return out;
+}
+
+GrpcPolledFdFactoryWindows::GrpcPolledFdFactoryWindows(IOCP* iocp)
+    : iocp_(iocp) {}
+
+void GrpcPolledFdFactoryWindows::Initialize(grpc_core::Mutex* mutex,
+                                            EventEngine* event_engine) {
+  mu_ = mutex;
+  event_engine_ = event_engine;
+}
+
+GrpcPolledFdReturnType GrpcPolledFdFactoryWindows::NewGrpcPolledFdLocked(
+    ares_socket_t as) {
+  auto it = sockets_.find(as);
+  GPR_ASSERT(it != sockets_.end());
+  return it->second;
+}
+
+void GrpcPolledFdFactoryWindows::ConfigureAresChannelLocked(
+    ares_channel channel) {
+  static const struct ares_socket_functions kCustomSockFuncs = {
+      &GrpcPolledFdFactoryWindows::Socket /* socket */,
```

Comment:
There is compile-time checking if you use the syntax `/*socket=*/&GrpcPolledFdFactoryWindows::Socket`

Reply:
Done.